File: Atividade Avaliativa Final/Javascript/entity/torneio/Torneio.js
export class Torneio {
  constructor(nome, times = []) {
    if (new.target === Torneio) {
      throw new TypeError("Cannot construct Torneio instances directly");
    }
    this.nome = nome;
    this.times = times;
  }

  adicionarTime(time) {
    if (time.modalidade !== this.getModalidade()) {
      throw new Error(`Este torneio é de ${this.getModalidade()}, não aceita times de ${time.esporte}`);
    }
    this.times.push(time);
  }

  gerarMataMata() {
    if (this.times.length < 2) {
      console.log("É necessário pelo menos 2 times para gerar mata-mata.\n");
      return [];
    }
  
    const timesCopia = [...this.times];
    if (timesCopia.length % 2 !== 0) {
      timesCopia.push({ nome: "BYE", jogadores: [], modalidade: this.getModalidade() });
    }
  
    const random = timesCopia.sort(() => Math.random() - 0.5);
  
    const confrontos = [];
    for (let i = 0; i < random.length; i += 2) {
      confrontos.push({
        time1: random[i].nome,
        time2: random[i + 1].nome
      });
    }
  
    return confrontos;
  }
  
  exibirMataMata() {
    const confrontos = this.gerarMataMata();
    if (confrontos.length === 0) return;
  
    const qtdTimes = confrontos.length * 2;
    const nomeFase = this._nomeFasePorQuantidadeTimes(qtdTimes);
    console.log(`\n=== ${nomeFase} ===`);
    confrontos.forEach((jogo, i) => {
      console.log(`Jogo ${i + 1}: ${jogo.time1} vs ${jogo.time2}`);
    });
  }
  
  _nomeFasePorQuantidadeTimes(qtd) {
    const fases = {
      16: "Oitavas de final",
      8: "Quartas de final",
      4: "Semifinal",
      2: "Final"
    };
    return fases[qtd] || `Fase eliminatória com ${qtd} times`;
  }
}

File: Atividade Avaliativa Final/Python/entity/torneio/torneio.py


from abc import ABC, abstractmethod
from typing import List
from entity.time import Time

class Torneio(ABC):
    def __init__(self, nome: str):
        self.nome = nome
        self.times: List[Time] = []

    def adicionar_time(self, time: Time):
        if time.modalidade != self.get_modalidade():
            raise ValueError(f"Este torneio é de {self.get_modalidade()}, não aceita times de {time.modalidade}")
        self.times.append(time)

    def gerar_mata_mata(self) -> List[tuple]:
        if len(self.times) < 2:
            print("É necessário pelo menos 2 times para gerar mata-mata.")
            return []

        times = self.times[:]
        if len(times) % 2 != 0:
            times.append(Time("BYE", self.get_modalidade()))

        import random
        random.shuffle(times)

        return [(times[i].nome, times[i+1].nome) for i in range(0, len(times), 2)]

    def exibir_mata_mata(self):
        confrontos = self.gerar_mata_mata()
        qtd_times = len(confrontos) * 2
        fase = self._nome_fase_por_quantidade_times(qtd_times)
        print(f"\n=== {fase} ===")
        for i, (t1, t2) in enumerate(confrontos, 1):
            print(f"Jogo {i}: {t1} vs {t2}")

    def _nome_fase_por_quantidade_times(self, qtd: int) -> str:
        fases = {
            16: "Oitavas de final",
            8: "Quartas de final",
            4: "Semifinal",
            2: "Final"
        }
        return fases.get(qtd, f"Fase eliminatória com {qtd} times")

    @abstractmethod
    def get_modalidade(self) -> str:
        pass

    @abstractmethod
    def get_posicoes(self) -> List[str]:
        pass